lib/predicate/registry.js

```diff
@@ -0,0 +1,367 @@
+        Records: [{
+            eventVersion: '2.0',
+            eventSource: 'aws:s3',
+            awsRegion: runParams.region,
+            eventTime: new Date(),
+            eventName: runParams.eventName,
+            userIdentity: {
+                principalId: runParams.userId,
+            },
+            requestParameters: {
+                sourceIPAddress: runParams.sourceIPAddress,
+            },
+            responseElements: {
+                'x-amz-request-id': runParams.amzRequestId,
+                'x-amz-id-2': runParams.amzRequestId2,
+            },
+            s3: {
+                s3SchemaVersion: '1.0',
+                configurationId: '00000000-0000-0000-0000-000000000000',
+                bucket: {
+                    name: runParams.bucket,
+                    ownerIdentity: {
+                        principalId: runParams.bucketOwnerId,
+                    },
+                    arn: runParams.bucketArn,
+                },
+                object: {
+                    key: runParams.key,
+                    size: runParams.size,
+                    eTag: runParams.etag,
+                    contentType: runParams.contentType,
+                    metadata: runParams.metadata,
+                    data: buf,
+                    versionId: 'N/A',
+                    sequencer: 'N/A',
+                },
+            },
+        }],
+    };
+}
+
+function tryRequire(p) {
+    const out = {};
+    try {
+        const fn = require(path.resolve(__dirname, p));
+        if (typeof fn !== 'function') {
+            throw new TypeError('Required object not function');
+        }
+        out.fn = fn;
+    } catch (e) {
+        out.err = e;
+    }
+    return out;
+}
+
+function getPredicate(registry, runParams, callback) {
+    const ve =
+        validate(runParamsSpec, runParams, new PredicateError(), runParams);
+    if (ve.hasValues()) {
+        return callback(ve);
+    }
+
+    const key = createKey(runParams);
+    const preds = registry.predicates[key];
+
+    if (!preds) {
+        return callback(ve.push(`No predicate exists for "${key}"`));
+    }
+
+    const k = runParams.key;
+    const klen = k.length;
+    let maxp = -1;
+    let fn = null;
+
+    // Linear search - for now we don't anticipate
+    // having too many prefixes for the same bucket.
+    Object.keys(preds).forEach(p => {
+        const plen = p.length;
+        if (plen > klen) {
+            return;
+        }
+        if (plen > maxp && k.indexOf(p) === 0) {
+            maxp = plen;
+            fn = preds[p];
+        }
+    });
+
+    if (!fn) {
+        return callback(ve.push(`No predicate exists for "${key}"`));
+    }
+    return callback(null, fn);
+}
+
+function runWrapped(context, callback) {
+    const { passthrough, event, fn } = context;
+    // NOTE: Domains are used to prevent exceptions from
+    // killing the worker process (which is not possible with
+    // promises). See tests for details.
```

Comment:
I don't like it either, but if the user predicate throws without a domain wrapping it, the exception will bubble up to the implicit domain created by the `cluster` module and the process will terminate. Maybe that's not a problem as long as the client's connection isn't dropped.